src/query/storages/common/cache/src/providers/table_data_cache.rs

```diff
@@ -0,0 +1,196 @@
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+use crate::metrics_inc_cache_access_count;
+use crate::metrics_inc_cache_hit_count;
+use crate::metrics_inc_cache_miss_count;
+use crate::metrics_inc_cache_population_overflow_count;
+use crate::metrics_inc_cache_population_pending_count;
+use crate::CacheAccessor;
+use crate::DiskBytesCache;
+use crate::DiskCacheBuilder;
+
+struct CacheItem {
+    key: String,
+    value: Arc<Vec<u8>>,
+}
+
+#[derive(Clone)]
+pub struct TableDataColumnCacheKey {
+    cache_key: String,
+}
+
+impl TableDataColumnCacheKey {
+    pub fn new(block_path: &str, column_id: u32) -> Self {
+        Self {
+            cache_key: format!("{block_path}-{column_id}"),
+        }
+    }
+}
+
+impl From<TableDataColumnCacheKey> for String {
+    fn from(value: TableDataColumnCacheKey) -> Self {
+        value.cache_key
+    }
+}
+
+impl AsRef<str> for TableDataColumnCacheKey {
+    fn as_ref(&self) -> &str {
+        &self.cache_key
+    }
+}
+
+/// Tiered cache which consist of
+/// - a bounded channel that keep the references of items being cached
+/// - a disk or redis based external cache
+#[derive(Clone)]
+pub struct TableDataCache<T = DiskBytesCache> {
+    external_cache: T,
+    population_queue: crossbeam_channel::Sender<CacheItem>,
+    _cache_populator: DiskCachePopulator,
+}
+
+const TABLE_DATA_CACHE_NAME: &str = "table_data_cache";
+
+pub struct TableDataCacheBuilder;
+impl TableDataCacheBuilder {
+    pub fn new_table_data_disk_cache(
+        path: &str,
+        population_queue_size: u32,
+        disk_cache_bytes_size: u64,
+    ) -> Result<TableDataCache<DiskBytesCache>> {
+        let disk_cache = DiskCacheBuilder::new_disk_cache(path, disk_cache_bytes_size)?;
+        let (rx, tx) = crossbeam_channel::bounded(population_queue_size as usize);
+        let num_population_thread = 1;
+        Ok(TableDataCache {
+            external_cache: disk_cache.clone(),
+            population_queue: rx,
+            _cache_populator: DiskCachePopulator::new(tx, disk_cache, num_population_thread)?,
+        })
+    }
+}
+
+impl CacheAccessor<String, Vec<u8>, DefaultHashBuilder, Count> for TableDataCache {
+    fn get<Q: AsRef<str>>(&self, k: Q) -> Option<Arc<Vec<u8>>> {
+        metrics_inc_cache_access_count(1, TABLE_DATA_CACHE_NAME);
+        let k = k.as_ref();
+        if let Some(item) = self.external_cache.get(k) {
+            metrics_inc_cache_hit_count(1, TABLE_DATA_CACHE_NAME);
+            Some(item)
+        } else {
+            metrics_inc_cache_miss_count(1, TABLE_DATA_CACHE_NAME);
+            None
+        }
+    }
+
+    fn put(&self, k: String, v: Arc<Vec<u8>>) {
+        // check if external(disk/redis) already have it.
+        if !self.external_cache.contains_key(&k) {
+            // populate the cache to external cache(disk/redis) asyncly
+            let msg = CacheItem { key: k, value: v };
+            match self.population_queue.try_send(msg) {
```

Comment:
Use `is_full` checking instead try to send and get the error?
https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Sender.html#method.is_full

Reply:
that maybe too conservative? `try_send` will put item into queue if possible and return `TrySendError::Full` if queue `is_full` during sending